### ISSUE TITLE
Epic bug: Please support my favorite package manager!

### ISSUE THREAD

**Issue by bitspittle** [varabyte/kobweb-cli#11]
In order to avoid everyone creating one off bugs for the same package manager, this bug will act as a home to collect all such requests received so far. (Of course, once someone starts working on adding support for one of them, that could be a good time to create an individual bug).

Here is the current list of package managers we are considering (but not actively working on):

* [apt](https://en.wikipedia.org/wiki/APT_(software))
* [chocolatey](https://chocolatey.org/)
* [flatpak](https://flatpak.org/)
* [gofish](https://gofi.sh/)
* [macports](https://www.macports.org/)
* [rpm](https://rpm.org/)
* [snapcraft](https://snapcraft.io/)
* [spec](https://rpm-packaging-guide.github.io/#what-is-a-spec-file)
* [winget](https://github.com/microsoft/winget-cli)

---

**NOTE**: **Some of these package managers may never get supported**, as I'm not sure how many people actually want to use them, and supporting more releases results in more work on our end.

So even if you see your package manager on this list, it still may be worth leaving a comment below like "Yes, I want that one please! I use it particularly because..."

If you see a comment below from someone expressing support for a package manager that you also want, please give their comment a thumbs up so we can better roughly understand what people are using.

**Comment by russellbanks** [comment]
It would be nice to have this on [Winget](https://github.com/microsoft/winget-cli)! Currently however, [Winget doesn't accept zip files](https://github.com/microsoft/winget-cli/issues/140) but rather installers so we'd have to either wait until they do support zip files (it's an open issue) or create an executable (exe or msi).

**Comment by bitspittle** [comment]
Thank you @russellbanks for your comment. I updated the initial list to include winget.

Hmmm, it might be possible to create an exe... but at that point I think I would be bundling a jre which might balloon the final size a bit, which may be unfortunate since to use Kobweb the user pretty has to have Java installed on their machine already. 

Hopefully, though, winget decides to support .zip files. It is so convenient and would reduce so much friction!

**Comment by russellbanks** [comment]
I'm not the most well-versed in packaging JVM applications (in terms of bundling the JRE, etc) but is it possible to create an executable without bundling the JRE? If this is not feasible, we could just bite the bullet and bundle the JRE with it, specifying that the recommended option is still the zip file?

Even if this is a temporary solution, I still think I would be advantageous to look into ways we can get an executable working in order to get this on Winget as it is the official package manager for Windows.

P.S. Just a correction to the updated initial list - the link for Winget is [https://github.com/microsoft/winget-cli](https://github.com/microsoft/winget-cli).

**Comment by bitspittle** [comment]
@russellbanks Thanks for letting me know about the wrong link. I fixed the original comment.

Even if we could build an .exe, just keep in mind I'd want a flow where this could all be done automatically every time I released a new version. With existing package managers, this is basically a one click operation for all of them so far. So we'd need to figure out how to automatically 1) unzip the file, 2) bundle it up into an exe, 3) and publish this to winget.

I personally won't have time to look into this for quite a while, but if someone knows how to do it, and/or has an example of this happening already, please add a comment -- that could help a lot!

P.S. Russell, you may want to check out the https://github.com/jreleaser/jreleaser project and file an issue with them to support winget. I'm relying on them pretty heavily for my other releases, and if winget is the official package manager for windows, they may be interested in getting some signal that people would want it supported.

**Comment by russellbanks** [comment]
Not much more I can comment on this other than I hope that either Winget supports zip files in the future or JReleaser adds support for Winget. There is an [open issue](https://github.com/jreleaser/jreleaser/issues/185) on JReleaser already to ask for support for winget as a packager, but it doesn't seem to have any activity.

**Comment by bitspittle** [comment]
Great, thanks for the update! With JReleaser leading the way, I will
definitely take a look.

**Comment by russellbanks** [comment]
As of winget [1.4](https://github.com/microsoft/winget-cli/releases/tag/v1.4.2161-preview), there will be support for installing zip files. However, this is just a preview at the moment, with the stable version being 1.3. But, this means that Kobweb can be put on winget; it's just a matter of time.

**Comment by aalmiray** [comment]
FWIW you can use `jpackage` (on its own or via JReleaser) to create Windows installers (exe, msi) that may be used in combination with Winget.

**Comment by bitspittle** [comment]
(Sorry for the recent radio silence, folks. I just moved, which wiped out my schedule for a couple of weeks).

As I catch up on stuff, I do intend to look into Winget, although I have a few other priorities to get through first. I'm hoping to get to this by end of September ('22) at the very latest, so feel free to ping this bug if it's later with no progress.

**Comment by bitspittle** [comment]
*Finally*, I'm back to this bug and looking into adding winget support. Thanks again for everyone's patience.

@aalmiray I'm trying to follow the Gradle syntax for jpackage from here: https://jreleaser.org/guide/latest/reference/assemble/jpackage.html

but "jpackage" contents aren't autocompleting for me:
![jreleaser-jpackage](https://user-images.githubusercontent.com/43705986/224174064-0c3d3e99-7002-4d40-80d7-bb54f28f460b.png)

(I'm using latest, Jreleaser v1.5.0)

I'm using kotlin script BTW. Maybe "app" is some dynamic thing that works only in groovy scripts?

Is there any demo code for using jreleaser + jpackage I can look at? 

Also, according to @russellbanks , winget should support .zip files now ([1.4](https://github.com/microsoft/winget-cli/releases/tag/v1.4.10173) was released in Jan). Are there plans to add support for that into JReleaser, or will `jpackage`ing be the way to go for a while?

**Comment by russellbanks** [comment]
@bitspittle

https://github.com/jreleaser/jreleaser/issues/185

https://jreleaser.org/guide/latest/reference/packagers/winget.html#_distribution_support

**Comment by bitspittle** [comment]
Right, that's what I'm looking at. `winget` doesn't support .zip file targets according to that distribution support list, so I assume at this point I still need to try running `jpackage` to create an .exe for now.

**Comment by aalmiray** [comment]
Zip support in Winget was added in `1.4` but the team has decided to wait until a critical mass of users has migrated to `1.4`. See https://github.com/jreleaser/jreleaser/issues/185#issuecomment-1406908541

Once that happens we'll support releasing zips with Winget.

re: auto-completion. `app` is a free name which you can choose to be anythign you want. It identifies a named element inside a NamedContainer, thus you must use regular KotlinDSL such as

```kotlin
jreleaser {
  assemble {
    jpackage {
      create("kobweb-installer") {
        ...
      }
    }
  }
}
```

**Comment by bitspittle** [comment]
Ah perfect, it was a Kotlin script thing. Thanks for the quick reply!

**Comment by russellbanks** [comment]
> Zip support in Winget was added in `1.4` but the team has decided to wait until a critical mass of users has migrated to `1.4`. See [jreleaser/jreleaser#185 (comment)](https://github.com/jreleaser/jreleaser/issues/185#issuecomment-1406908541)

@aalmiray zip's were turned on in the pipelines a month or so ago:
https://github.com/microsoft/winget-pkgs/issues/87300

**Comment by aalmiray** [comment]
@russellbanks gotcha. Will have a look at zips on Winget in the next couple of weeks.

**Comment by aalmiray** [comment]
Filed a new ticket to keep track of zip support on Winget https://github.com/jreleaser/jreleaser/issues/1253

**Comment by bitspittle** [comment]
I think I'll hold off on winget until zip support is ready.

It looks like supporting building an MSI/EXE installer will be rather tricky, involving a bunch of structure just for winget that no other package manager has needed. For example, to run `jlink / jpackage`, I think I would need to set up a special CI workflow using Java 16, since everything else I've done works fine with JDK11 (or older). 

And given the extra complexity around this, I have other priorities I should attend to first.

**Comment by russellbanks** [comment]
I can manually submit kobweb to winget for the time being if you'd like @bitspittle

**Comment by aalmiray** [comment]
Understood. If it helps at a later stage, have a look a JReleaser's [release workflow](https://github.com/jreleaser/jreleaser/blob/main/.github/workflows/release.yml) and [configuration](https://github.com/jreleaser/jreleaser/blob/main/jreleaser.yml) which take care of multiple targets during a release.

**Comment by bitspittle** [comment]
Thanks @russellbanks , I appreciate the offer!

However, I'd want to make sure that everything is handled in the build script so whenever I update the Kobweb binary, I can do everything with a single `jreleaserPublish`.

Probably best to figure everything out the correct way all at once (or otherwise I might release future versions of Kobweb and winget would fall behind)

**Comment by russellbanks** [comment]
Anything you do do to automate winget should take the previous manifest and update the values with that of your new artifacts, so a pre-existing manifest shouldn't stop you doing that in the future.

**Comment by bitspittle** [comment]
@aalmiray Yeah I was using one of your simple workflows as a guide, but seeing that I'd need  to set up a bunch of runtime JDKs just for winget seems like a lot of work.

Also, I can just jlink / jpackage a windows binary right? If I don't care about the other platforms?

And if I am building an installer for windows, do I need to make sure to run jreleaser on a Windows machine?

**Comment by bitspittle** [comment]
@russellbanks where does the winget manifest live? Currently, I have some local repositories (e.g. for homebrew) which jreleaser overwrites for me. Is winget different? (I was skimming the docs but haven't looked too closely yet)

**Comment by aalmiray** [comment]
> Also, I can just jlink / jpackage a windows binary right? If I don't care about the other platforms?

Correct.

> And if I am building an installer for windows, do I need to make sure to run jreleaser on a Windows machine?

Correct, again.

**Comment by russellbanks** [comment]
> @russellbanks where does the winget manifest live? Currently, I have some local repositories (e.g. for homebrew) which jreleaser overwrites for me. Is winget different? (I was skimming the docs but haven't looked too closely yet)

[microsoft/winget-pkgs](https://github.com/microsoft/winget-pkgs)

Pull requests are just made to winget-pkgs, usually using and created with a tool like [Komac](https://github.com/russellbanks/Komac), [YamlCreate](https://github.com/microsoft/winget-pkgs/blob/master/Tools/YamlCreate.ps1) or [WinGetCreate](https://github.com/microsoft/winget-create).

P.S. I'm a moderator for winget-pkgs so you can ask me any questions

**Comment by bitspittle** [comment]
@russellbanks Ah thanks, really helpful info! I don't think I'd mind you submitting an initial manifest for me, then, if it doesn't block me from updating or overwriting it later.

I was trying to search randomly through the manifests.... do you know of a manifest out there which uses the new zip type?

FYI latest Kobweb zip is here: https://github.com/varabyte/kobweb/releases/tag/cli-v0.9.11

However, it's not a standalone zip file. The user will need to have Java 11 (or newer) installed on their machine. 

Does winget support the idea of specifying "Java 11" as a dependency, at which point it handles installing things automatically? Or would it be up to me to bundle Java 11 into the zip file?

@aalmiray Does this mean when I run `jreleaser` targeting winget, it won't immediately update anything but will submit a PR on my behalf?

**Comment by russellbanks** [comment]
Ah, winget doesn't support dependencies currently. It is planned as a "soon" feature. See [Dependencies.md on winget-cli](https://github.com/microsoft/winget-cli/blob/73f05cdf9b72eed991ccc5609f459f385cb9ec2f/doc/specs/%23163%20-%20Dependencies.md).

For the time being, you could make a second zip that has a stripped JRE using JLink bundled with it? I believe that's what Compose does when you run `createDistributable` (My [compose project](https://github.com/russellbanks/HashHash) is on desktop so may just be for that). That's entirely up to you though.

**Comment by aalmiray** [comment]
> @aalmiray Does this mean when I run jreleaser targeting winget, it won't immediately update anything but will submit a PR on my behalf?

Not by default, no. The original idea was to push to an intermediate repository (`winget-kobweb` for example) and trigger a GH workflow that pull create a branch and push. The PR would still have to be done manually _unless_ the workflow were to use the `gh` command. This is how https://github.com/jreleaser/jreleaser-macports is handled.

However, I just added an option to specify a `branchPush` property on taps/repositories (https://github.com/jreleaser/jreleaser/issues/1129) which would give you the ability to directly create a branch on a target repository such as `winget-pkgs`, bypassing `winget-kobweb`. The actual PR would still have to be created manually at the moment.

**Comment by bitspittle** [comment]
Hmmm, I'm thinking I should probably should look into jlink at this point. 

Give me a week or so, I have some other things I need to focus on first, but I'll ramp up on jlink / jpackage over the next few days.

Maybe I'll even give in and finally move publishing my artifacts via Jreleaser from a CI instead of my personal desktop, like an actual professional or something :stuck_out_tongue: Once that's migrated, then it wouldn't be too hard to add a workflow that runs on Windows for Winget.

(I just spent a week getting CI artifact support working for two of my other projects. It's a pain, to be honest, and I'm not looking forward to the game of CI trial and error yet again at the moment)

@russellbanks No `createDistributable` or anything like it in my tasks, sadly. Note that it seems pretty unlikely for Kobweb users to *not* have Java installed on their machine, since they're doing Kotlin development anyway. Having winget handle the dependencies for me would be the best though, I think.

**Comment by aalmiray** [comment]
> (I just spent a week getting CI artifact support working for two of my other projects. It's a pain, to be honest, and I'm not looking forward to the game of CI trial and error yet again at the moment)

That's why JReleaser has a `dry-run` mode and offers plenty of options to run as much as possible on a local environment. Alas, when platform specific binaries or tools are required you must test on the given platform.

**Comment by skaldebane** [comment]
Hi there! It would be nice to support `.rpm` packages for Fedora and other RHEL-based Linux distros. Thanks!

**Comment by skaldebane** [comment]
Flatpak is also a pretty solid one. It ships with almost every Linux distro by default (barring Ubuntu), and Flathub allows publishing CLI apps (although they won't be shown in the apps store understandably).

**Comment by bitspittle** [comment]
@Skaldebane thanks! I updated the original comment.

Do you have any experience with either of those package managers? At one point, I checked out snap and it was a wasted day.

**Comment by skaldebane** [comment]
@bitspittle I've previously created an .rpm for a specific purpose, but nothing really proper.
As for Flatpak, I do have some experience with it (for making a GTK app), but not in the Java ecosystem.

However, I'll try to package Kobweb into one of them when I find some time!

Edit: Both are apparently supported by JReleaser.

**Comment by aalmiray** [comment]
FWIW JReleaser supports both packagers.

The Native Installer assembler relies on Jpackage to create an .rpm that bundles a Java Runtime along with the app. 

The Spec packager OTOH builds an .rpm as expected by Fedora COPR’s build infrastructure. 

Lastly, Flatpak may be used to package CLI and/or desktop apps

**Comment by bitspittle** [comment]
(Ah sorry @aalmiray meant to reply last week!)

Just wanted to say thanks for the last comment. If we can do everything with JReleaser, that's definitely the way to go.

@Skaldebane if you were going to try to experiment with this yourself, the jreleaser config / kobweb binary is actually in a different project. [You can find it here](https://github.com/varabyte/kobweb-cli/blob/2be5ac3b2eeb90ca2ca9c735f5b8cc75ce53dd28/kobweb/build.gradle.kts#L73). No pressure if you don't have time to work on it; if you ever do, feel free to reach out to me for any help you need.

**Comment by Ayfri** [comment]
Any update on winget support?

**Comment by bitspittle** [comment]
Thanks for the reminder @Ayfri . I added it to my near-term todo list to revisit, to see if it will work seamlessly now with latest JReleaser.

I probably have a week or so more of Kobweb work to get through, and then I'll see if I can get winget support working after that. I'll report back here whether it goes well or if there are still obstacles.

**Comment by bitspittle** [comment]
> I probably have a week or so more of Kobweb work to get through, and then I'll see if I can get winget support working after that. I'll report back here whether it goes well or if there are still obstacles.

Just checking in that I've been slammed by a handful of surprises after posting this comment, so my week estimate has apparently ballooned into a month. However, I haven't forgotten this issue, and I am hoping to take a look into it later this week or next.

**Comment by Ayfri** [comment]
Haha I see, no problem, unforeseen events happen to anyone !

**Comment by bitspittle** [comment]
Just an update to my winget folks, I do not think the attempt to support winget could be going more poorly.

I've got a bunch of other tasks on my plate, so I'm going to abandon this for now. Apologies at it seems like this is the most requested package manager at the moment. 

I will enumerate my concerns below. People with more knowledge of the matter can certainly point out flaws in my thinking or outright mistakes.

* jreleaser doesn't seem to support a `Dependencies` block (@aalmiray am I missing something here?). It seems like I would probably want to add a `PackageDependencies` entry for `Microsoft.OpenJDK.11` in my case.
  * Note that I briefly looked into jlink but it doesn't seem trivial and I really don't have many hours right now to chase it down. If someone is willing to add jlink support to `kobweb-cli` that *might* help unblock winget support.
* jreleaser seems to drop its code into a root directory when I *believe* should be dropping it under a `manifests` directory in accordance with the winget PR requested format.
  * e.g. [`v/Varabyte/kobweb/0.9.13`](https://github.com/varabyte/winget-kobweb/tree/main/v/Varabyte/kobweb/0.9.13) should probably be `manifests/v/varabyte/kobweb/0.9.13`?
  * @aalmiray instructions in the docs walking you through uploading a PR for a winget project could be extremely beneficial IMO.
  * Also @aalmiray you mentioned something about using `branchPush` to push directly to the winget-pkdgs repo but that sounds scary and could also potentially use some docs.
* At this point, winget would be the first package manager that would require me to open an update PR each time I release a new version of Kobweb, which is a HUGE point of friction. 
  * Normally, after I release a new version of Kobweb, I update [a file](https://github.com/varabyte/data/blob/main/kobweb/cli-version.txt) which will send a message to Kobweb CLI users that a new version is ready. With winget, I'd have to wait for the PR process to finish before tweaking this file. Otherwise, windows users would be getting a "You should upgrade!" message before they can actually upgrade.
  * What if users report a bug with a new version that I have to fix quickly? If I'm blocked by a winget PR, I can't notify people that a new version is out.
  * Unless I could get a guarantee from winget-pkgs folks that PRs for version updates were automatic and I had a 100% guarantee that I'd get approved on the order of minutes, this lack of control is a dead end to me.
* When I tested installing the zip file of my Kobweb CLI using a local copy of a winget manifest, winget gave me a big scary red error that malware was detected in my zip file. I had to override it to proceed, but.... won't users see that error? The last thing I need when releasing Kobweb software is windows users piling on to tell me my software has malware in it.
* Once I finally installed kobweb locally using winget, it didn't even run. I got an error message popup: "Cannot start or run due to incompatibility with 64-bit versions of Windows. Please contact the software vendor to ask if a 64-bit Windows compatible version is available." and pressing OK caused windows to flash. 

---

**My recommendation at this point for Windows folks is to just use `scoop` or `sdkman`.**

If winget would consider supporting external manifests added manually via remote repositories, and if it wouldn't incorrectly report my code has malware, AND if I could figure out why I was getting those weird "This software is not compatible with 64-bit windows", then maybe I could be convinced to revisit winget support in the near future.

**Comment by aalmiray** [comment]
@bitspittle I've raised https://github.com/jreleaser/jreleaser/issues/1596 and https://github.com/jreleaser/jreleaser/issues/1597 for the first 2 items.

Using `branchPush` allows you to create a branch other than the current one when pushing to a given repository. JReleaser lets you by default to push to a winget package repository of your own. It would be your responsibility to create a PR to the shared package repository on your own.

By using the `branchPush` you may be able (if access rights have been granted) to create such branch directly on the shared package repository. Take note that JReleaser does not yet provides the means to create PRs automatically. Even if a branch were to be created in this way you'd still have to manually create the PR.

The only advantage of using `branchPush` in this way is to skip your own package repository.

**Comment by aalmiray** [comment]
One more thing, the report as malware is likely caused by an unsigned package. https://github.com/jreleaser/jreleaser/issues/689 keeps track of this issue

**Comment by bitspittle** [comment]
Thank you, as always @aalmiray , for the insanely fast response and turnaround time.

If the malware error comes from the lack of signing, it's a shame that winget isn't taking the 256 sha into account. I mean, I'm essentially confirming that my artifacts are hosted in my own repo. I know that just because you're hosting open source software on github, that doesn't mean it can't have malware, but then again, I could always sign my own software which has malware in it, so it's not like it provides that much of an extra benefit as far as I can tell.

That said, I have set up signing for maven in a bunch of places, so if it was clear to me exactly what file winget wanted me to sign and where that key should be found, I could look into adding it, assuming it's some standard signing technology and not something that is windows only. (I have `gpg` set up in other projects of mine)

**Comment by aalmiray** [comment]
😅 Thanks. It was fun adding what was missing for winget. 

Regarding signing, if this is indeed the case then I'm afraid you must use Windows' `signtool` which is setup way differently to PGP (what Maven Central requires at the moment).

**Comment by EthanRStokes** [comment]
Would you consider supporting an AUR package?

There *is* a package on the AUR, but it's very out of date. The latest supported version is 0.9.12.

**Comment by aalmiray** [comment]
@bitspittle you may recall we have pending support for AUR in JReleaser https://github.com/jreleaser/jreleaser/issues/693

**Comment by bitspittle** [comment]
@aksh1618 are you still around? The AUR work you did for Kobweb so many years ago broke when I had to move the kobweb CLI out into a different repository. Would it be easy to fix up the script you wrote to use this new repository location?

**Comment by aksh1618** [comment]
Hey @bitspittle! Ah I see, I think the action should work in this repository as well without too many changes, let me double check and get it back up this weekend. I'll reach out to you on discord if I need anything w.r.t. permissions etc.

**Comment by bitspittle** [comment]
@aksh1618 I found that you actually shared the private key with me many years ago which I forgot I had. Looking over the old github workflows, I believe I have everything I need to update the AUR package on my end. Thanks!! I'll see if I can get the AUR package updated in the next day or two.

**Comment by bitspittle** [comment]
@Treetrain1 would you mind trying now to see if AUR is correctly hooked up again? (@aksh1618 you might be able to check as well but no pressure)

Here are some old instructions, which I'll add back into the Kobweb README if this is confirmed working.

--- 
## Arch Linux

### With an AUR helper:

```bash
$ trizen -S kobweb
```

### Without an AUR helper:

```bash
$ git clone https://aur.archlinux.org/kobweb.git
$ cd kobweb
$ makepkg -si
```

**Comment by aksh1618** [comment]
@bitspittle I just checked it on my machine, seems to be working perfectly! The instructions also look good :+1:

**Comment by bitspittle** [comment]
Thank you @aksh1618 for confirming!

By the way, when a new version of Kobweb comes out, if you already have a previous version installed, what is the proper way to update? 

(I'm asking because I need to update [this section](https://github.com/varabyte/kobweb#update-the-kobweb-binary))

**Comment by aksh1618** [comment]
In this case there is no separate command to update, it will basically be the same as the installation command. So maybe you can put the same command. (Usually on arch you periodically run updates using the aur helper and everything that has an update gets updated in one go)

**Comment by bitspittle** [comment]
@Treetrain1 thanks for leaving a comment! Arch Linux should be supported now (again). At this point, moving forward, the AUR kobweb entry should get automatic updates again.

Instructions added into the README here: https://github.com/varabyte/kobweb#arch-linux

@aalmiray thanks for considering AUR support in jreleaser. At the moment, I don't need it, and I definitely don't expect you to prioritize it, but if it ever makes it in, let me know, and I can migrate over to it. Feel free to reach out to me if you want me to share all the steps I'm currently doing at the moment to automate an AUR package publication.

**Comment by bitspittle** [comment]
Uh whoops, I have no recollection of closing this issue (which I mean to keep open forever basically). I must have misclicked on mobile. Maybe I should move it to a discussion someday instead. Anyway, reopening.

**Comment by lilazero** [comment]
winget is basically becoming the goto for windows so i think it's a must

**Comment by lheintzmann1** [comment]
Hello 👋 
It would be nice to have Nix/NixOS support 😁

**Comment by bitspittle** [comment]
> Hello 👋 It would be nice to have Nix/NixOS support 😁

@lheintzmann1 I haven't tried Nix myself but I've heard great things. Can you tell me what Nix users expect when it comes to having support to install some binary? What's the process?

**Comment by lheintzmann1** [comment]
@bitspittle Sure ! Nix/NixOS users typically expect a reproducible, declarative, and isolated package installation process.
I suggest you to have a look at [the official packages repo](https://github.com/NixOS/nixpkgs), you will need to add a package there.
Let me know if you'd like some help !

**Comment by bitspittle** [comment]
@lheintzmann1 thanks for the offer of help! 

One critical requirement for me to add support for a target at this point, at least pre-1.0, is full control. _At the moment_, I'm not comfortable relying on a third party GitHub repo (even one as official as NixOS) because I don't own the pipeline. 

For an example of why this might matter: when a new update goes out, I increment a [version file](https://github.com/varabyte/data/blob/main/kobweb/cli-version.txt) that CLIs ping occasionally to see if they need to update. At the moment, this would be terrible for NixOS users, because they'd be told to grab latest but when they try to do that it won't work (my request to update official packages will be stuck behind a code review, if I'm understanding the process correctly)

Does NixOS support unofficial package repositories that users can manually extend their OS to check? Like how [Arch Linux](https://kobweb.varabyte.com/docs/getting-started/getting-kobweb#arch-linux) lets people specify a special Kobweb target, or how the other package managers I support require a one time step that sets them up to integrate my Kobweb specific installation steps.

**Comment by dshatz** [comment]
I have made a simple Nix flake for kobweb-cli for private use. Feel free to grab. I will accept PRs for version updates.
https://github.com/dshatz/kobweb-cli-nix

P.S. it is not building from source, but grabbing from github release page.

**Comment by bitspittle** [comment]
@dshatz would it be trivial (and if so, would I have your permission) to fork that over into a project that I own under the varabyte repo? (Basically, if I add support for Nix, [I'll also add an additional rule to my workflow file](https://github.com/varabyte/kobweb-cli/blob/20f337d358937089f4a14f6fc4fa2b9cfc341a0d/.github/workflows/publish.yml#L74)). Basically, I would want to tweak my Nix flake to update the version and sha when I publish a new CLI.

(BTW, speaking of, new CLI coming out soon, hopefully mid-next week)

**Comment by BierDav** [comment]
Thumbs up for `winget` because it is simply the native solution Windows machines

**Comment by Grigoriym** [comment]
Thumbs up for `apt`

**Comment by AshyPinguin** [comment]
can there be a gentoo ebuild for kobweb? (`portage`)

**Comment by dshatz** [comment]
> [@dshatz](https://github.com/dshatz) would it be trivial (and if so, would I have your permission) to fork that over into a project that I own under the varabyte repo? (Basically, if I add support for Nix, [I'll also add an additional rule to my workflow file](https://github.com/varabyte/kobweb-cli/blob/20f337d358937089f4a14f6fc4fa2b9cfc341a0d/.github/workflows/publish.yml#L74)). Basically, I would want to tweak my Nix flake to update the version and sha when I publish a new CLI.
> 
> (BTW, speaking of, new CLI coming out soon, hopefully mid-next week)

Sorry, of course you have my permission!

**Comment by bitspittle** [comment]
* Thanks @dshatz , no promise on timelines but I'll see if I can port it over at some point. I should probably create a Nix VM at some point as well; at the moment, I am aware of the power of Nix but, not having tried it yet, I'm flying a bit blind :)

* I'd love to support `winget` but every time in the past I checked, I hit a blocker. It always felt like I just needed to let some time pass. I'll try again soon around the same time I play around with `Nix`.

* I recall adding myself as a package to `apt` was really intimidating. I will probably wait for a 1.0 release before retrying that.

For Nix, winget, and even apt, if there is anyone here who is an expert in any of them and could possibly help me expedite this, feel free to reach out to me directly. My email is listed on [my github profile page](https://github.com/bitspittle).

**Comment by AshyPinguin** [comment]
> can there be a gentoo ebuild for kobweb? (`portage`)

I made an ebuild that is currently in the dev branch of guru repo https://github.com/gentoo/guru/commit/bbfeb48a8a955a6089311a70808ab4f28d7a97d9

I might in the future try to make a source build.

**Comment by bitspittle** [comment]
> > can there be a gentoo ebuild for kobweb? (`portage`)
> 
> I made an ebuild that is currently in the dev branch of guru repo [gentoo/guru@bbfeb48](https://github.com/gentoo/guru/commit/bbfeb48a8a955a6089311a70808ab4f28d7a97d9)
> 
> I might in the future try to make a source build.

Hello @AshyPinguin I'm very sorry I didn't reply to your comment before. I did not mean to ignore it.

In general, my main requirement for a new package manager at this point are that is has a decent number of users and, more important, that I can run an action on the CI that automatically publishes a new version of the CLI without me having to do anything manually. I'm not too familiar with gentoo so I'll take a look in a bit. I'm currently on vacation through June 4 so I've added a reminder to tackle this after I've gotten back and taken a few days to recover from jetlag.

I really appreciate people taking their own time to make contributions to benefit my projects, so no matter what, thank you very much for writing the code.

**Comment by AshyPinguin** [comment]
> Hello [@AshyPinguin](https://github.com/AshyPinguin) I'm very sorry I didn't reply to your comment before. I did not mean to ignore it.

Don't worry, it is fine.

> In general, my main requirement for a new package manager at this point are that is has a decent number of users and, more important, that I can run an action on the CI that automatically publishes a new version of the CLI without me having to do anything manually. I'm not too familiar with gentoo so I'll take a look in a bit.

It currently uses the release artifacts so it doesn't need a CI job or something (I also don't think it would be possible to automate generating an ebuild for it in CI with the fact that that means pushing directly to the guru repo which I wouldn't think the developers overlooking it will like.). If I ever make a source ebuild sometime, I'll say it when it is handy to have a CI job to publish for example depencies as a tarball.

> I'm currently on vacation through June 4 so I've added a reminder to tackle this after I've gotten back and taken a few days to recover from jetlag.

Enjoy your vacation, no need to rush. :)

> I really appreciate people taking their own time to make contributions to benefit my projects, so no matter what, thank you very much for writing the code.

No problem :p I have a bit too much time on my hands and personally don't like needing multiple package managers for different packages so this also fixes that.

**Comment by bitspittle** [comment]
Heads up all that Nix support for the Kobweb CLI is coming soon, and technically here now, thanks to user @e-psi-lon. He did a bunch of work and sent me his project today, so I've just been coming up to speed and getting ready to integrate it into the CLI publish flow.

It will probably be another day or two of tinkering and updating official docs, but if you're eager to try it now, you can visit https://github.com/varabyte/kobweb-cli-nix and follow the instructions in the README.

I am _not_ a Nix user myself, so if anyone has any feedback about the README or the experience, feel free to let me know.
